src/rmc.rs

```diff
@@ -55,164 +42,58 @@ pub struct OutgoingMessage<H: Signable, MK: MultiKeychain> {
-        }
-    }
-
-    fn handle_unchecked_message(
-        maybe_state: Option<Self>,
-        message: IncomingMessage<H, MK>,
-        node_count: NodeCount,
-        keychain: &'a MK,
-        multisigned_hashes_tx: &UnboundedSender<Multisigned<'a, H, MK>>,
-    ) -> Option<Self> {
-        match message.message {
-            Message::SignedHash(unchecked) => {
-                let signed_indexed_hash = match unchecked.check(keychain) {
-                    Ok(signed_indexed_hash) => signed_indexed_hash,
+            _ => match message {
+                Message::MultisignedHash(unchecked) => match unchecked.check_multi(keychain) {
+                    Ok(complete) => Some(HashState::Completed { complete }),
                     Err(_) => {
-                        debug!("Received a hash with a bad signature");
-                        return maybe_state;
+                        debug!(target: "rmc", "Received a hash with a bad signature");
+                        maybe_state
                     }
-                };
-                let indexed_hash = signed_indexed_hash.as_signable();
-                let index = indexed_hash.index();
-                if index != message.sender {
-                    debug!("Index of a message does not match the sender");
-                    // This is an unusual situation, but we can still process the hash because
-                    // it is signed by the node with the index from the message.
-                }
-                Self::handle_signed_hash(
-                    maybe_state,
-                    signed_indexed_hash,
-                    node_count,
-                    keychain,
-                    multisigned_hashes_tx,
-                )
-            }
-            Message::MultisignedHash(unchecked) => {
-                let signed_indexed_unchecked = match unchecked.check(keychain) {
-                    Ok(signed_indexed_unchecked) => signed_indexed_unchecked,
-                    Err(_) => {
-                        debug!("Received a multisigned hash with a bad signature");
-                        return maybe_state;
+                },
+                Message::SignedHash(unchecked) => {
+                    let signed_hash = match unchecked.check(keychain) {
+                        Ok(signed_hash) => signed_hash,
+                        Err(_) => {
+                            debug!(target: "rmc", "Received a hash with a bad multisignature");
+                            return maybe_state;
+                        }
+                    };
+                    let partial = match maybe_state {
+                        None => signed_hash.into_partially_multisigned(keychain),
+                        Some(HashState::SignatureAggregation { mut partial }) => {
+                            partial.add_signature(signed_hash);
+                            partial
+                        }
+                        Some(HashState::Completed { .. }) => unreachable!(),
```

Comment:
Ouch, now I see why you complained. How about abstracting this, together with the state change code below, into a function again, and explicitly handling this case there (by returning the unchanged state from the function ofc)?

Reply:
I refactored the code completely, and replaced the HashState type with PartiallyMultiSigned. The message handling logic is now simple enough that I decided to move it to a single method of ReliableMulticast. This allows to avoid passing a ton of arguments to a function and writing a spaghetti code